blog/2024-11-mastering-effect-and-afterRenderEffect/README.md

```diff
@@ -0,0 +1,477 @@
+  In this example no cleanup is required, but we wanted to mention the fact that long-running tasks (such as a timeout) should be cleaned up. The cleanup will be called before entering the same phase again, or if the effect itself is destroyed.
+  The `write` phase then passes the new height, `newHeight`, to the `read` phase.
+  Hint: Pass the same value to `read` (eg. `return 100`) and you will see that the follow-up phase won't be executed.
+  Setting the same number twice won’t be considered a change, so the `write` phase won’t mark the `read` phase as dirty.
+
+3. **read Phase**: The final `read` phase logs the `newHeight`. 
+  We could also read from the DOM in that phase and store the result to a new signal. But in this example this work is not necessary, because the `earlyRead` is already doing that job.
+
+> We encourage you to scroll down to check out our Demo Application. 
+  Feel free to follow the hints in the comments to experiment with the specifics of each phase.
+
+## Migration Guide: From Angular Lifecycle Hooks to Signal-Based Reactivity
+
+With Angular 19, the Angular team's broader vision of signal-based components is slowly taking shape.
+The long-term goal here is to eventually phase out all traditional lifecycle hooks, except for `ngOnInit` and `ngOnDestroy`.
+
+The addition of `effect()` and `afterRenderEffect()` showcases how Angular is moving in this direction. 
+These effects are more intuitive for managing component state changes and post-render interactions, thus making the old lifecycle hooks redundant.
+For instance, `afterRenderEffect()` is designed to handle tasks traditionally managed by `ngAfterViewInit` and `ngAfterViewChecked`.
+
+This approach has been in the pipeline for some time. 
+In April 2023, Angular’s team outlined this trajectory in their [RFC #49682](https://github.com/angular/angular/discussions/49682). 
+The document proposed the introduction of `afterRenderEffect()` as part of a roadmap to replace Angular’s current change detection, moving away from imperative lifecycle hooks to a cleaner, more reactive pattern. 
+
+Migrating from Angular lifecycle hooks to `effect()` and `afterRenderEffect()` is straightforward:
+- **ngOnInit / ngOnChanges** → `effect()`: Handles signal-based logic and other state.
+- **ngAfterViewInit / ngAfterViewChecked** → `afterRenderEffect()`: Manages DOM manipulations post-render.
+
+Or to put it another way, here’s a direct mapping:
+
+| Lifecycle Hook        | Replacement            |
+|-----------------------|------------------------|
+| `ngOnInit`            | `effect()`             |
+| `ngOnChanges`         | `effect()`             |
+| `ngAfterViewInit`     | `afterRenderEffect()`  |
+| `ngAfterViewChecked`  | `afterRenderEffect()`  |
+
+Now the only important hook left is actually `ngOnDestroy`.
+Will we also get a replacement for this, or will we no longer need this functionality once everything has been completely migrated to signals? 🤔
```

Comment:
Isn't [`DestroyRef`](https://angular.dev/api/core/DestroyRef) what could replace `ngOnDestroy`?

Reply:
DestroyRef is how `takeUntilDestroyed` works internally 🙂 it's been there since Angular 16